petastorm/pytorch.py

```diff
@@ -0,0 +1,118 @@
+                       result in an infinite number of epochs.
+        :param sequence: If it is set to a Sequence object, then fetch will return a sequence, otherwise fetch
+                       will return an item.
+        :param training_partition: An int denoting the partition number used for multi node training. Each node should
+                       pass in a unique partition number in the range [0, num_training_partitions).
+                       num_training_partitions must be supplied as well.
+        :param num_training_partitions An int denoting the number of training partitions (how many nodes are performing
+                       the multi node training)
+        :param read_timeout_s: A numeric with the amount of time in seconds you would like to give a read before it
+                       times out and raises an EmptyResultError. Pass in None for an infinite timeout
+        :param cache: An object conforming to `cache.CacheBase` interface. Before loading row groups from a parquet file
+                       the Reader will attempt to load these values from cache. Caching is useful when communication
+                       to the main data store is either slow or expensive and the local machine has large enough storage
+                       to store entire dataset (or a partition of a dataset if num_training_partitions is used).
+        """
+        self.batch_size = batch_size
+        self.collate_fn = collate_fn
+        self.num_epochs = num_epochs
+        self.transform = transform
+
+        if num_workers == 0:
+            # Use 1 thread
+            num_workers = 1
+        if shuffle_options == True:
+            shuffle_options = ShuffleOptions()
+        self.reader = Reader(dataset_url, schema_fields=schema_fields,
+                             predicate=predicate, rowgroup_selector=rowgroup_selector,
+                             reader_pool=ThreadPool(num_workers), num_epochs=num_epochs, sequence=sequence,
+                             training_partition=training_partition, num_training_partitions=num_training_partitions,
+                             read_timeout_s=read_timeout_s, cache=cache, shuffle_options=shuffle_options)
+
+    def __len__(self):
+        return (len(self.reader) + self.batch_size - 1) // self.batch_size
+
+    def __iter__(self):
+        """
+        The Data Loader iterator stops the for-loop at the end of each epoch, but a subsequent for-loop
+        will instantiate a new Reader and yield more results, until the requested number of epoch has been
+        reached.  After that point, any subsequent call results in StopIteration, per iterator protocol.
+        """
+        batch = []
+        for row in self.reader:
+            batch.append(self.transform(row))
+            if len(batch) == self.batch_size:
+                yield self.collate_fn(batch)
+                batch = []
+        if batch:
```

Comment:
pytorch's `DataLoader` has `drop_last` to define whether we want an incomplete batch. Should be simple to add the implementation here. Ok if you want as a separate CL.

Reply:
I think that makes sense. Let me add that in a separate CL/PR.